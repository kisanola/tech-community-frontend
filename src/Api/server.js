import axios from 'axios';

const token = localStorage.getItem('token') || null;
const server = axios.create({
  baseURL: process.env.REACT_APP_BACKEND_URL,
  headers: {
    Authorization: token,
    'Content-Type': 'application/json',
  },
});

server.interceptors.request.use(
  async config => {
    const token = await localStorage.getItem('token');
    if (token) config.headers['Authorization'] = token;
    return config;
  },
  error => {
    return Promise.reject(error);
  },
);

export default server;
