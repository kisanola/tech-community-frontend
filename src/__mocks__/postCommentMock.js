export default {
  _id: 1,
  author: { picture: 'img.jpg', username: 'username' },
  createdAt: '2019-10-07',
  body: 'Hello world',
};
