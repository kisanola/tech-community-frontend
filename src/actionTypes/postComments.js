export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const SET_COMMENTS = 'SET_COMMENTS';
