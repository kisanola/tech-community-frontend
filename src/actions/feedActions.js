import server from '../Api/server';

import * as types from '../actionTypes/feed';

import { FEED_LIMIT } from '../constants';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOOGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const setFeed = (payload) => ({
  type: types.SET_FEED,
  payload,
});

/**
 * Set feed organization
 * @return {object} action
 */
export const setFeedOrganizations = (payload) => ({
  type: types.SET_FEED_ORGANIZATIONS,
  payload,
});

/**
 * Get feed
 * @return {object} response
 */
export const getFeed = ({
  limit,
  itemsLength,
  category = 'general',
  search = '',
}) => async (dispatch) => {
  if (limit > itemsLength + FEED_LIMIT) return;

  dispatch(toggleLoading(true));

  try {
    const {
      data: { feed },
    } = await server.get(
      `/api/v1/feed?offset=${itemsLength}&limit=${FEED_LIMIT}&category=${category}&search=${search}`,
    );

    dispatch(setFeed(feed));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};


/**
 * Get feed
 * @return {object} response
 */
export const getFeedOrganizations = () => async (dispatch) => {
  const res = await server.get(
    '/api/v1/organizations?limit=5',
  );
  if (res) {
    const {
      data: { organizations },
    } = res;
    dispatch(setFeedOrganizations(organizations.docs || []));
  }
};
