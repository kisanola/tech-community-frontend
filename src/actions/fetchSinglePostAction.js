import server from '../Api/server';

import * as types from '../actionTypes/singlePostTypes';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOGGLE_LOADING,
  payload: state,
});

/**
 * Set feed
 * @return {object} action
 */
export const fetchSuccess = (post) => ({
  type: types.FETCH_SUCCESS,
  payload: post,
});

/**
 * Get feed
 * @return {object} response
 */
export default (slug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    const {
      data: { post },
    } = await server.get(
      `/posts/${slug}`,
    );

    dispatch(fetchSuccess(post));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
