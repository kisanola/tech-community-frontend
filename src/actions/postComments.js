import server from '../Api/server';
import * as types from '../actionTypes/postComments';

/**
 * Toggles the loading state
 * @return {object} action
 */
export const toggleLoading = (state) => ({
  type: types.TOGGLE_LOADING,
  payload: state,
});

/**
 * Set comments
 * @return {object} action
 */
export const setComments = (comments) => ({
  type: types.SET_COMMENTS,
  payload: comments,
});

/**
 * Get comments
 * @return {object} response
 */
export const getPostComments = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    const {
      data: { postComments },
    } = await server.get(`/posts/${postSlug}/Comments`);

    dispatch(setComments(postComments));
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
