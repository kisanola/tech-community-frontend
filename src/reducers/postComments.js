import * as types from '../actionTypes/postComments';

const reducer = (state = { loading: true }, { type, payload }) => {
  switch (type) {
    case types.TOGGLE_LOADING:
      return {
        ...state,
        loading: payload,
      };
    case types.SET_COMMENTS:
      return {
        ...state,
        items: payload,
      };
    default:
      return state;
  }
};

export default reducer;
