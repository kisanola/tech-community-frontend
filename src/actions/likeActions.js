import server from '../Api/server';

import { toggleLoading } from './feedActions';

/**
 * Like a post
 * @return {object} action
 */
export const like = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.post(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};

/**
 * Unlike a post
 * @return {object} action
 */
export const unlike = (postSlug) => async (dispatch) => {
  dispatch(toggleLoading(true));

  try {
    await server.delete(`/posts/${postSlug}/like`);
    dispatch(toggleLoading(false));
  } catch (e) {
    dispatch(toggleLoading(false));
  }
};
