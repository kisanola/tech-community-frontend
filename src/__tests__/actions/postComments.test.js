import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import server from '../../Api/server';
import { getPostComments } from '../../actions/postComments';
import postCommentMock from '../../__mocks__/postCommentMock';

const mockStore = configureMockStore([thunk]);

describe('Login actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({});
  });

  test('Should call the `getPostComments` action with success', () => {
    jest.spyOn(server, 'get').mockResolvedValue({
      data: { postComments: [postCommentMock] },
    });

    store.dispatch(getPostComments('slug')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });

  test('Should call the `getPostComments` action with success', () => {
    jest.spyOn(server, 'get').mockRejectedValue({
      error: {},
    });

    store.dispatch(getPostComments('slug')).then(() => {
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
