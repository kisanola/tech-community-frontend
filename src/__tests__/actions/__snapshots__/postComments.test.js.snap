// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login actions Should call the \`getPostComments\` action with success 1`] = `
Array [
  Object {
    "payload": true,
    "type": "TOGGLE_LOADING",
  },
  Object {
    "payload": Array [
      Object {
        "_id": 1,
        "author": Object {
          "picture": "img.jpg",
          "username": "username",
        },
        "body": "Hello world",
        "createdAt": "2019-10-07",
      },
    ],
    "type": "SET_COMMENTS",
  },
  Object {
    "payload": false,
    "type": "TOGGLE_LOADING",
  },
]
`;

exports[`Login actions Should call the \`getPostComments\` action with success 2`] = `
Array [
  Object {
    "payload": true,
    "type": "TOGGLE_LOADING",
  },
  Object {
    "payload": false,
    "type": "TOGGLE_LOADING",
  },
]
`;
